### PR TITLE
[8.x] Add ability to dispatch unique jobs (Option 1)

### PR DESCRIPTION
**Pull request by paras-malhotra**
See also https://github.com/laravel/framework/pull/35042 (alternative)

### Motivation:
This PR allows the ability to dispatch unique jobs. So, if a unique job is already dispatched, it will not dispatch another job (with the same key) unless the job has completed processing. See https://github.com/laravel/ideas/issues/2151.
Ping: @themsaid and @royduin

### Difference between this PR and `WithoutOverlapping`:
The difference between `WithoutOverlapping` and dispatching unique jobs is that the `WithoutOverlapping` middleware processes jobs uniquely (meaning two jobs with the same key will not be processed in parallel). However, it does not stop jobs (with the same key) from being enqueued if another such job is already in the queue waiting to be processed.

### Use Case:
This can be quite useful in scenarios such as search indexing. If a search index job is already enqueued for a resource, you probably don't want another such job to be queued until the previous job has finished processing. Sidekiq Enterprise offers the [unique jobs](https://github.com/mperham/sidekiq/wiki/Ent-Unique-Jobs) feature as well.

### How to Use
It's incredibly easy to use this. If you wish to dispatch a unique job, just chain the `unique` method like so:

```php
// Unique for the model ID and timeout is set at 1 hour
MyUniqueJob::dispatch($model)->unique($model->id, 3600);

// Can also do this
dispatch(new MyUniqueJob($model))->unique($model->id, 3600);

// Global unique job with a 1 hour timeout
MyUniqueJob::dispatch(...)->unique();
```

### CODE REVIEW

File: src/Illuminate/Queue/Middleware/ReleaseUniqueJobLock.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+class ReleaseUniqueJobLock
+{
+    /**
+     * Process the job.
+     *
+     * @param  mixed  $job
+     * @param  callable  $next
+     * @param  string  $key
+     * @return mixed
```
Comment (driesvints) [inline src/Illuminate/Queue/Middleware/ReleaseUniqueJobLock.php:16]:
```suggestion
     * @return void
```

File: src/Illuminate/Foundation/Bus/PendingDispatch.php [hunk 2]
```diff
@@ -142,7 +175,9 @@ public function __call($method, $parameters)
      */
     public function __destruct()
     {
-        if ($this->afterResponse) {
+        if($this->cancelDispatch) {
```
Comment (driesvints) [inline src/Illuminate/Foundation/Bus/PendingDispatch.php:178]:
```suggestion
        if ($this->cancelDispatch) {
```

File: src/Illuminate/Foundation/Bus/PendingDispatch.php [hunk 3]
```diff
@@ -109,6 +120,28 @@ public function chain($chain)
         return $this;
     }
 
+    /**
+     * Indicate that the job should be dispatched as a unique job.
+     *
+     * @param string  $uniqueBy
+     * @param int  $uniqueFor
```
Comment (driesvints) [inline src/Illuminate/Foundation/Bus/PendingDispatch.php:127]:
```suggestion
     * @param  string  $uniqueBy
     * @param  int  $uniqueFor
```

**Comment by paras-malhotra** [comment]
Thanks @driesvints, I've updated the formatting based on your comments. 👍

**Comment by themsaid** [comment]
I think it'd be simpler if you set a unique param on the job and use it to create the lock. Also you can release the lock inside `CallQueuedHandler` when a job succeeds or fails. That way you won't need to use a middleware.

**Comment by paras-malhotra** [comment]
> I think it'd be simpler if you set a unique param on the job and use it to create the lock.

@themsaid, I thought about that but it would be a breaking change if users already have that param set on the job. Definitely seems cleaner though.

> you can release the lock inside `CallQueuedHandler`

I didn't want to mess with the handler code, but if you think that's a better approach, I can update the PR to do that.

**Comment by themsaid** [comment]
You may call it something like `$JobDeduplicationId` so there's less chance it'll conflict with an existing property.

As for the handler, IMO it's fine to modify it for this feature.

**Comment by paras-malhotra** [comment]
I've opened a new PR with a different implementation/syntax as discussed.

**Pull request closed without merging**